core/src/block.rs

```diff
@@ -389,91 +513,84 @@ mod valid {
-                    self,
-                    SignatureVerificationError::SignatureMissing.into(),
-                )));
-            }
-
-            if !self.0.replace_signatures(signatures) {
-                return WithEvents::new(Err((
-                    self,
-                    SignatureVerificationError::UnknownSignature.into(),
-                )));
-            }
+        ) -> Result<(), iroha_crypto::Error> {
+            let signatory = &topology.as_ref()[signature.0 as usize];
+            self.0.add_signature(signature, signatory.public_key())
+        }
 
-            let actual_block_hash = self.as_ref().hash();
-            if actual_block_hash != expected_hash {
-                let err = BlockValidationError::IncorrectHash {
-                    expected: expected_hash,
-                    actual: actual_block_hash,
-                };
+        /// Replace block's signatures. Returns previous block signatures
+        ///
+        /// # Errors
+        ///
+        /// - Replacement signatures don't contain the leader signature
+        /// - Replacement signatures contain duplicate signatures
+        /// - Replacement signatures contain unknown signatories
+        /// - Replacement signatures contain incorrect signatures
+        pub fn replace_signatures(
+            &mut self,
+            signatures: Vec<BlockSignature>,
+            topology: &Topology,
+        ) -> WithEvents<Result<Vec<BlockSignature>, SignatureVerificationError>> {
+            let prev_signatures = self.0.replace_signatures_unchecked(signatures);
 
-                return WithEvents::new(Err((self, err)));
+            if let Err(err) = Self::verify_leader_signature(self.as_ref(), topology)
+                .and_then(|()| Self::verify_validator_signatures(self.as_ref(), topology))
+            {
+                self.0.replace_signatures_unchecked(prev_signatures);
+                WithEvents::new(Err(err))
+            } else {
+                WithEvents::new(Ok(prev_signatures))
             }
-
-            self.commit(topology)
         }
 
-        /// Verify signatures and commit block to the store.
+        /// commit block to the store.
         ///
         /// # Errors
         ///
-        /// - Not enough signatures
-        /// - Not signed by proxy tail
+        /// - Block has duplicate proxy tail signatures
+        /// - Block is not signed by the proxy tail
+        /// - Block doesn't have enough signatures
         pub fn commit(
             self,
             topology: &Topology,
         ) -> WithEvents<Result<CommittedBlock, (ValidBlock, BlockValidationError)>> {
-            if !self.0.header().is_genesis() {
-                if let Err(err) = self.verify_signatures(topology) {
+            if !self.as_ref().header().is_genesis() {
+                if let Err(err) = Self::verify_proxy_tail_signature(self.as_ref(), topology) {
                     return WithEvents::new(Err((self, err.into())));
                 }
```

Comment:
We probably should check that whole set of sigantures is valid.

Reply:
it's checked on `replace_signatures`

Reply:
because `ValidBlock` is guaranteed to have all signatures valid